src/Xamarin.Android.Build.Tasks/Utilities/TypeMapGenerator.cs

```diff
@@ -491,6 +554,10 @@ void OutputModule (ModuleDebugData moduleData)
 		//
 		void OutputModule (BinaryWriter bw, ModuleDebugData moduleData)
 		{
+			if ((uint)moduleData.JavaToManagedMap.Count == UInt32.MaxValue) {
```

Comment:
This feels like dead/impossible code.  Sure, there's "defense in depth," but this feels like it's bordering on lunacy, as `List<T>` is backed by an array which can hold at most [`Array.MaxArrayLength`](https://github.com/dotnet/runtime/blob/0ec974709e311ef63e7e23ed7c74f1f33762aecc/src/libraries/System.Private.CoreLib/src/System/Array.cs#L27) elements.

This is admittedly less than `int.MaxValue`, and is an implementation detail.

Allowing `List<T>.Capacity` and `List<T>.Count` to *exceed* `int.MaxValue` would be a *yuge* semantic break, though.  Checking that `int.MaxValue` is less than `uint.MaxValue` is bizarre; it can't *not* be true.

(And if by some chance somebody *wanted* to allow `List<T>`/Arrays to hold more than int.MaxValue elements -- as mono could do at one point in time! -- you'd have to have `Count` *throw an exception* when Count exceeded `int.MaxValue`, otherwise currently semantically correct code that checks that `list.Count > 0` *will **fail***, which is very much not good.  Which means that *this* code would be *doubly* broken!)

If anything, use `int.MaxValue` as the limit, not `uint.MaxValue`.  Assuming we need this check at all.

Reply:
It's a check for an "impossible" situation, for sure. But I don't want to rely on implementation details of `List<>` to check a condition that may affect us on the runtime. Also, if for some weird reason (a bug?) `.Count` is set to `-1`, then casting it to `uint` will yield `uint.MaxValue`. However, that is not the point of all this - `uint.MaxValue` is a special value used by the native runtime to determine that the entry should be skipped (is considered as "invalid") and this check makes this condition more explicit while not hurting anything.